V2 version of SwiftUI ChartView

AppPear opened the pull request:
The goal is to have a new more easily expandable codebase, setup with CI and test so PRs could be merged quicker and safer.

**If you would like to help and participate in the new version you are very welcomed** you can find project related tasks [here](https://github.com/AppPear/ChartView/projects/2#column-9329701) or under the Issues with a `v2` tag

*to submit your work please fork this repo and create a new branch from `new-version` branch then update with your work and create a PR for the `new-version` branch.*

If you work on a ticket please go to the project board and assign the ticket to yourself and move it to the in-progress column. If you are finished please assign the newly created PR to me and move your ticket to the ready-for-review column.

Join our Slack channel for day to day conversation and more insights:

[slack invite link](https://join.slack.com/t/swiftuichartview/shared_invite/zt-el5pnmba-FvyraEsI~EwxqWHNfHZWZg)

New code proposal:

### ChartViews

These chart views come with the bare minimum for a chart, so it can be easily built into custom user interfaces. We also build on them for the more complex chart widgets.

*Data*
* should be a structure which can hold a `x` and `y` value. 
* `y` value must be numeric (possibly floating point number)
* `x` can be numeric or string 
*  has to be at least 3 initialisers: 1)  only `y` values, 2) `x` and `y` numeric values 3) x `string` and `y` numeric values

*Barchart:*

```swift
@State var chartData: [Double] = [0, 5, 6, 2, 13, 4, 3, 6]

BarChart()
.data(chartData)
.style(lightOrange)
```

*Linechart:*

```swift
@State var chartData: [Double] = [0, 5, 6, 2, 13, 4, 3, 6]

LineChart()
.data(chartData)
.style(lightOrange)
```

*Piechart:*

```swift
@State var chartData: [Double] = [0, 5, 6, 2, 13, 4, 3, 6]

PieChart()
.data(chartData)
.style(lightOrange)
```

*Grid* (not in beta 1)
* is a view under a chart 
* should support `barchart` and `linechart`
* legend ranges can be user set
* legend ranges can be automatic
* user can set a custom style (colors)

```swift
Grid {
    BarChart()
}
.data(chartData)
.style(lightOrange)
```

*Cardview*
* support for all charts
* a holder for the chart, title, subtitle
* support for displaying `x` axis values interactively 
* should be rounded
* light versions have a shadow around them
* dark versions have a thin light bezel
* have 3 size presets: `small`, `medium`, `large`
* user can set a custom style (colors)
* has a defualt style so setting a style is optional

```swift
CardView {
        BarChart()
}
.data(chartData)
.style(darkBlue)
```

### ChartData (beta 1 supports Chart init with [Double] only)

ChartData is an observable object so you can make state changes and it will be reflected now in the charts.

To initialise:

`ChartData(_ data: [Double])`

and to make a state change:

```swift
@State var chartData = ChartData([0, 4, 5])
.
.
.
chartData.data = [0, 2, 6, 7, 8, 11]
```

AppPear committed:
Replace the old with a more sleek and flexible code architecture

satan87 commented:
I believe this will break a lot of other PR. But looks cleaner, and easier so I don't mind adapting my PR.

LucasCarioca commented:
With this in the works, should people hold off on PRs? 

Do you want to open up issues to break up this change and get help implementing it?

AppPear committed:
Added a first implementation of BarChart and LineChart also introduced style

satan87 commented:
To participate in the new branch, how the best way to do it?
Just fork it, update it, then create a PR?

AppPear commented:
Yes, please fork it and create a PR for the `new-version` branch. 
You can find some task under the issues with a `v2` tag or you can go to the project board and find the tickets I created and would be very helpful if you could work on those.

Thanks!

satan87 commented:
ok, I'll try working on the PieChart, but no promises;(.

satan87 commented:
I created a first PR with minimum changes, to make sure I follow your guidelines.
Let me know if it's ok.
Maybe we could create chat somewhere to coordinate.

AdrianBinDC commented:
@AppPear I love the philosophy re: building the charts w/ the bare minimum and building upon it. One suggestion I have is writing unit tests to go along with committed code, rather than treating it as an afterthought. I'll take a peek at code coverage and see what I can unearth that's testable.

AdrianBinDC committed:
Write unit tests for CGPoint+Extension.swift (#100)

- write unit tests for CGPoint+Extension.swift
- clean up formatting on CGPoint+Extension.swift

L1cardo commented:
Looking forward to this！！
This will be awesome！

satan87 committed:
Add PieChart + multicolor (#98)

* Add ColorGradient example
Add ColorGradient single color constructor
Add preview for BarChart

* Add PieChart
Allow multi color for Pie and Bar
Add linter

AppPear committed:
removed .rotate for foreGroundColor

AppPear commented:
> ~I created a first PR with minimum changes, to make sure I follow your guidelines.~

Thanks, I merged it! Only small issue was `accentColor:self.style.foregroundColor.rotate(for: index)` that .rotate() functions was undefined so I had to remove it if you could do a fix PR defining it and reimplementing that would be awesome! @satan87

satan87 commented:
> > ~I created a first PR with minimum changes, to make sure I follow your guidelines.~
> 
> Thanks, I merged it! Only small issue was `accentColor:self.style.foregroundColor.rotate(for: index)` that .rotate() functions was undefined so I had to remove it if you could do a fix PR defining it and reimplementing that would be awesome! @satan87

The extension file wasn't in my commit;( sorry.
PR #102

satan87 committed:
Restore rotating index for multicolor (#102)

AdrianBinDC committed:
Write unit tests for Color+Extension.swift (#101)

- Wrote tests
- Possible minor bug where there's an alpha missing on 32-bit colors?

satan87 commented:
Hello,
Is there a way in V2 to show labels?
I see some code, but not sure how to add it to a chart?
Thanks,
Nicolas

PS:
As it seems there are a couple of persons interested in the project, maybe we could set up a Sclak or something similar

satan87 commented:
I spend some time trying to add the TITLE in the Chart View but I couldn't do it.
I didn't fully grasp the whole concept yet...
Nicolas

AppPear commented:
@satan87 slack channel is a great idea I updated the PRs description with the invite link.

Currently the TitleLabel and the ChartsLabel files are I think a dead end. I have a different solution in mind. 

I think would be better to have a CardView as a container view to hold the title and the chartview. CardView would have viewbuilder and would work something like this:

CardView {
    Title(“Text”)
     ChartView(...)
}

satan87 committed:
Syntax corrections (#105)

* Movie PieChartFile

* Variable renaming

AppPear committed:
added mac os as a build target

satan87 committed:
Bug Fix: Bar Chart with [0] crashed (#110)

AppPear committed:
Add CardView and CardLabel (#111)

AppPear committed:
Creating a data structure which propagets changes in data to the charts (#114)

* Creating a data structure which propagets changes in data to the charts

* Fixed appearing animation

AppPear committed:
Added ChartLabel interaction

AppPear committed:
Refactor Chart base (#143)

AppPear committed:
Add public modifier to ChartColors and add showShadow property

AppPear committed:
Fix cornerMasking on card view when no shadow is set

sagar5534 committed:
Dark/Light mode fixes (#148)

Fix for making text work with both Dark/Light mode.

Also solves line chart background to appear white in dark mode

LucasCarioca commented:
Tried out the Beta and it is looking pretty cool. How close is this to getting released?

danwood committed:
recalculate geometry if orientation has changed (#156)

Co-authored-by: Dan Wood <danwood@users.noreply.github.com>

danwood committed:
Issue 99 documentation (#159)

* Starting on filling in documentation.

* First pass on most/all files

* more descriptions filled in

* Some documentation but TBH the author would be better suited to explain how this works!

* more basic stuff filled in

* Add a description and bunch of discussion text for most of the view `body` declarations

* more explanations

Co-authored-by: Dan Wood <danwood@users.noreply.github.com>

danwood committed:
Activity-type Rings charts (#161)

Co-authored-by: Dan Wood <danwood@users.noreply.github.com>

AppPear committed:
V2 beta 2 changes (#150)

* Add PieChart interaction PR changes to v2

* Add custom string format for ChartLabel when interactionInProgress = true (#151)

* Dark/Light mode fixes (#148)

Fix for making text work with both Dark/Light mode.

Also solves line chart background to appear white in dark mode

* Add custom string format for ChartLabel when interactionInProgress = true

Co-authored-by: Sagar Patel <s.72427patel@gmail.com>

* Prepare charts to display x and y values souch as a value for a given point

Co-authored-by: Roddy Munro <roddymunro@icloud.com>
Co-authored-by: Sagar Patel <s.72427patel@gmail.com>

AppPear committed:
Add a public init() to RingsChart

shreyaspapi commented:
Waiting for the release😁

kakashilw commented:
In my widget i import SwiftUICharts. 2.0.0-beta.1，and my code

import SwiftUICharts

...............

struct TestWidgetView : View {
    
    var entry: TestEntry

    
    var body: some SwiftUI.View {
        
        let orangeStyle = ChartStyle(backgroundColor: .white,
                                     foregroundColor: [ColorGradient(ChartColors.orangeBright, ChartColors.orangeDark)])
        LineChart()
        .data([0, 5, 6, 2, 13, 4, 3, 6])
        .chartStyle(orangeStyle)

    }
    
    
}

and run，the errors：


Showing Recent Messages
Undefined symbol: nominal type descriptor for SwiftUICharts.ChartStyle

Undefined symbol: nominal type descriptor for SwiftUICharts.LineChart

Undefined symbol: nominal type descriptor for SwiftUICharts.ChartData

Undefined symbol: nominal type descriptor for SwiftUICharts.ChartValue

Undefined symbol: protocol conformance descriptor for SwiftUICharts.LineChart : SwiftUI.View in SwiftUICharts

Undefined symbol: SwiftUICharts.ColorGradient.init(SwiftUI.Color, SwiftUI.Color) -> SwiftUICharts.ColorGradient

Undefined symbol: SwiftUICharts.ChartColors.orangeDark.unsafeMutableAddressor : SwiftUI.Color

Undefined symbol: type metadata accessor for SwiftUICharts.ChartStyle

Undefined symbol: SwiftUICharts.LineChart.init() -> SwiftUICharts.LineChart

Undefined symbol: type metadata for SwiftUICharts.ColorGradient

Undefined symbol: SwiftUICharts.ChartColors.orangeBright.unsafeMutableAddressor : SwiftUI.Color

Undefined symbol: type metadata for SwiftUICharts.LineChart

Undefined symbol: protocol witness table for SwiftUICharts.LineChart : SwiftUICharts.ChartBase in SwiftUICharts

Undefined symbol: SwiftUICharts.ChartStyle.__allocating_init(backgroundColor: SwiftUI.Color, foregroundColor: [SwiftUICharts.ColorGradient]) -> SwiftUICharts.ChartStyle

Undefined symbol: (extension in SwiftUICharts):SwiftUI.View< where A: SwiftUICharts.ChartBase>.data([Swift.Double]) -> some

Undefined symbol: (extension in SwiftUICharts):SwiftUI.View.chartStyle(SwiftUICharts.ChartStyle) -> some

sgallo0692 commented:
@kakashilw isn't this line:

var body: some SwiftUI.View {

Supposed to be 

var body: some View {

champagnepoon commented:
hello @AppPear  i found i faile to add the package because of the package.resolved file which is out of date
what should i do

m8xp0w3r commented:
Hi,
are you still working on this? Last commits are from october 2020. 😢

kevinmun commented:
I actually tried importing this and using it in Widget, but it draws blank on both the preview and simulator. ANyone able to get it working?

m8xp0w3r commented:
@kevinmun yes, it works in my application.

kevinmun commented:
> @kevinmun yes, it works in my application.

I see, I tested it out again, and it seems, it's ok if it's deployed to device. Somehow on preview / simulator, it appears as blank. Weird.

JanC commented:
hey, is the v2 still being worked on or is this library not maintained anymore?

AppPear commented:
> hey, is the v2 still being worked on or is this library not maintained anymore?

Yes, currently I'm on orginising a bigger help to maintein and take this lib to next level and release v2 as soon as possible

m8xp0w3r commented:
Hi,
I've got a question to BarChart. Tthis is used with key value pairs (String -> Double). If I touch a single Bar in V1, both values are presented.
![image](https://user-images.githubusercontent.com/5719156/105627061-a1281100-5e34-11eb-87f4-3e8b6c13825e.jpeg)
But in V2 only the Double value is shown.
![image](https://user-images.githubusercontent.com/5719156/105626957-de3fd380-5e33-11eb-9380-fb6b9613c346.jpeg)
Is this a bug, a forgotten feature or wrong usage from my side?

Prince2k3 commented:
@AppPear I'd like to join as a member. I currently have a candlestick chart working using the new-version branch on my machine. I'm interested in helping out. I like to chat about data structure for support of that type of chart and others.

AppPear commented:
> @AppPear I'd like to join as a member. I currently have a candlestick chart working using the new-version branch on my machine. I'm interested in helping out. I like to chat about data structure for support of that type of chart and others.

Can you please join the slack channel?

tpgmeligmeyling commented:
@AppPear Can you update the Slack invite link?

AppPear commented:
> Hi,
> I've got a question to BarChart. Tthis is used with key value pairs (String -> Double). If I touch a single Bar in V1, both values are presented.
> ![image](https://user-images.githubusercontent.com/5719156/105627061-a1281100-5e34-11eb-87f4-3e8b6c13825e.jpeg)
> But in V2 only the Double value is shown.
> ![image](https://user-images.githubusercontent.com/5719156/105626957-de3fd380-5e33-11eb-9380-fb6b9613c346.jpeg)
> Is this a bug, a forgotten feature or wrong usage from my side?

it's a missing feature for now, but it is coming soon to v2 as well

m8xp0w3r commented:
> it's a missing feature for now, but it is coming soon to v2 as well

Thank for the answer.

AppPear committed:
feat(core): refactoring chart dispalying (#191)

now it is possible to add background lines precisely as charts are displayed at correct size
also rewrote basics to conform with Shapes and Animatable protocol

RSickenberg commented:
This project seems dead, unfortunately :(

rodri0315 commented:
Hello! How's V2 going? Anything we can help?

AppPear committed:
feat: add linechart interaction point (#202)

* feat: add linechart interaction point

* feat: add ability to show current data point on linechart

AppPear committed:
fix: BarChartCellShape to handle negative numbers correctly (#250)

AppPear committed:
fix: remove UIColors which caused CI build errors (#251)

AppPear committed:
feat: new protocol for chained functions, and added support for expli… (#252)

* feat: new protocol for chained functions, and added support for explicit Y ranges. X coming as well

* feat: add new axis interface (#253)

AppPear committed:
Feat/new protocol and range (#255)

* feat: new protocol for chained functions, and added support for explicit Y ranges. X coming as well

* feat: add new axis interface (#253)

AppPear committed:
feat: add animation toggle interface (#256)

Co-authored-by: “Andras <“andras.samu@supercharge.io”>

AppPear committed:
Refactor composable API and unify x-axis alignment

AppPear committed:
Ignore Xcode workspace user data for showcase app

AppPear committed:
Improve dark mode defaults and showcase styling

AppPear committed:
Add line chart drag interaction and showcase card

AppPear committed:
Add legend-driven series visibility controls

AppPear committed:
Add streaming data source with windowed auto-scroll

AppPear committed:
Add auto axis ticks with collision handling and rotation

AppPear committed:
Improve accessibility metadata and add high-contrast presets

AppPear committed:
feat: add chart performance mode with line downsampling

AppPear committed:
feat: add callback-based chart selection events

AppPear committed:
feat: add tab-based showcase app with dynamic data lab

champagnepoon commented:
您好，我是潘倩萍，我已经收到您的邮件，如果有问题我会尽快联系您的；如果有急事，烦请电话联系我，谢谢！
